openhtf/io/frontend/src/app/utils.ts

```diff
@@ -0,0 +1,252 @@
+  };
+}
+
+
+/** A Pipe for turning status codes into materialize-css color classes.**/
+@Pipe({name: 'statusToColor'})
+export class StatusToColor extends MapPipe {
+  map = {
+    'RUNNING': 'blue lighten-1 white-text',
+    'ONLINE': 'blue lighten-1 white-text',
+    'PASS': 'green lighten-1 white-text',
+    'FAIL': 'red lighten-1 white-text',
+    'ERROR': 'yellow darken-3 white-text',
+    'PENDING': 'blue-grey lighten-3 white-text',
+    'WAITING': 'blue-grey lighten-3 white-text',
+    'OFFLINE': 'blue-grey lighten-3 white-text',
+  };
+}
+
+
+/** A Pipe for darkening status colors. **/
+@Pipe({name: 'statusToDark'})
+export class StatusToDark extends MapPipe {
+  map = {
+    'RUNNING': 'blue darken-2 white-text',
+    'ONLINE': 'blue darken-2 white-text',
+    'PASS': 'green darken-2 white-text',
+    'FAIL': 'red darken-2 white-text',
+    'ERROR': 'yellow darken-4 white-text',
+    'PENDING': 'blue-grey lighten-2 white-text',
+    'WAITING': 'blue-grey lighten-2 white-text',
+    'OFFLINE': 'blue-grey lighten-2 white-text',
+  };
+}
+
+
+/** A Pipe for darkening status colors even more. **/
+@Pipe({ name: 'statusToSuperDark' })
+export class StatusToSuperDark extends MapPipe {
+  map = {
+    'RUNNING': 'blue darken-4 white-text',
+    'ONLINE': 'blue darken-4 white-text',
+    'PASS': 'green darken-4 white-text',
+    'FAIL': 'red darken-4 white-text',
+    'ERROR': 'yellow darken-5 white-text',
+    'PENDING': 'blue-grey lighten-1 white-text',
+    'WAITING': 'blue-grey lighten-1 white-text',
+    'OFFLINE': 'blue-grey lighten-1 white-text',
+  };
+}
+
+
+/** A Pipe for turning outcome codes into materialize-css color classes.**/
+@Pipe({ name: 'outcomeToColor' })
+export class OutcomeToColor extends MapPipe {
+  map = {
+    'PASS': 'green-text text-lighten-4',
+    'FAIL': 'red-text text-lighten-4',
+    'UNSET': 'blue-grey-text text-lighten-4'
+  };
+}
+
+
+/** A Pipe for colorizing log entries based on logging level. **/
+@Pipe({name: 'loggingLevelToColor'})
+export class LoggingLevelToColor extends MapPipe {
+  map = {
+    0: 'blue-text text-lighten-1',   // NOTSET
+    10: 'blue-text text-lighten-1',   // DEBUG
+    20: 'blue-text text-lighten-1',   // INFO
+    30: 'yellow-text text-darken-4',  // WARNING
+    40: 'red-text text-lighten-1',    // ERROR
+    50: 'red-text text-lighten-1'     // CRITICAL
+  };
+}
+
+
+/** A pipe for turning an Object into something Angular2 can loop over. **/
+@Pipe({name: 'objectToArray', pure: false})
+export class ObjectToArray implements PipeTransform {
+  result: any[];
+
+  transform(input: any): any {
+    this.result = [];
+    if (input == undefined || input == null) {
+      return this.result
+    }
+    let keys = Object.keys(input);
+    keys.forEach(key => {
+      this.result.push({ 'key': key, 'value': input[key] });
+    });
+    return this.result;
+  }
+}
+
+
+/** A pipe to recurisely stringify objects into html <ul> elements. **/
+@Pipe({name: 'objectToUl'})
```

Comment:
this pipe may suffer from the same strange issue if it isn't marked as non-pure, since it also works with composite objects.